File: Sorting.java
/*
 * @author ryan scherbarth 
 * @date 2/14/24
 */
public class Sorting {

  public static void main(String[] args) {
    // arrays to be sorted
    int[] arr1 = { 10, 198, 15, 78, 88, 61, 0, 17, 9 };
    int[] arr2 = { 68, 28, 10, 84, 74, 45, 41, 2, 31, 85, 43, 26, 45, 61, 4, 74, 22, 54, 74, 15, 47, 58, 58, 18, 61, 100, 89 };
    int[] arr3 = { 75, 51, 44, 86 };
    int[] arr4 = { 78, 37, 3, 14, 83, 32, 30, 73, 5, 81, 37, 69, 95 };
    int[] arr5 = { 27, 37, 49, 5, 45, 63, 51, 68, 5, 18, 84, 87, 40, 55, 58, 100, 90, 2, 61, 1, 89, 73, 60, 5, 92, 54, 32, 73, 36, 86, 10, 44, 82, 99, 67, 85, 14, 95, 32, 76, 47, 87, 48, 47, 75, 69 };

    // insertion
    System.out.println("------ Insertion Sort ------");
    testInsertionSort(arr1);
    testInsertionSort(arr2);
    testInsertionSort(arr3);
    testInsertionSort(arr4);
    testInsertionSort(arr5);
    System.out.println("--------------------------");

    // merge
    System.out.println("--------Merge Sort--------");
    testMergeSort(arr1);
    testMergeSort(arr2);
    testMergeSort(arr3);
    testMergeSort(arr4);
    testMergeSort(arr5);
    System.out.println("--------------------------");
  }

  public static void testInsertionSort(int[] in) {
    int cp[] = new int[in.length];
    for (int i = 0; i < in.length - 1; i++) {
      cp[i] = in[i];
    }
    int[] sorted = insertionSort(cp);

    System.out.print("Array before sorting: ");
    printArray(in);

    System.out.print("\nArray after sorting: ");
    printArray(sorted);

    System.out.println();
    System.out.println();
  }

  public static void testMergeSort(int[] in) {
    int cp[] = new int[in.length];
    for (int i = 0; i < in.length - 1; i++) {
      cp[i] = in[i];
    }
    int[] sorted = mergeSort(cp, 0, cp.length - 1);

    System.out.print("Array before sorting: ");
    printArray(in);

    System.out.print("\nArray after sorting: ");
    printArray(sorted);

    System.out.println();
    System.out.println();
  }

  public static void printArray(int[] arr) {
    for (int i = 0; i < arr.length; ++i) {
      System.out.print(arr[i] + " ");
    }
  }

  /*
   * Sorting Things
   */

  public static int[] insertionSort(int[] in) { // p.51
    int n = in.length;
    for (int i = 1; i < n; ++i) { // index 2..n
      int key = in[i];
      int j = i - 1;

      while (j >= 0 && in[j] > key) {
        in[j + 1] = in[j];
        j = j - 1;
      }
      in[j + 1] = key;
    }
    return in;
  }

  public static int[] mergeSort(int[] in, int p, int r) {
    if (p < r) {
      int q = (p + r) / 2;
      mergeSort(in, p, q);
      mergeSort(in, q + 1, r);
      merge(in, p, q, r);
    }
    return in;
  }

  public static void merge(int[] in, int p, int q, int r) {
    int n1 = q - p + 1;
    int n2 = r - q;
    int[] L = new int[n1];
    int[] R = new int[n2];

    for (int i = 0; i < n1; ++i) {
      L[i] = in[p + i];
    }
    for (int j = 0; j < n2; ++j) {
      R[j] = in[q + j + 1];
    }

    int i = 0, j = 0;
    int k = p;
    while (i < n1 && j < n2) {
      if (L[i] <= R[j]) {
        in[k] = L[i];
        i++;
      } else {
        in[k] = R[j];
        j++;
      }
      k++;
    }

    while (i < n1) {
      in[k] = L[i];
      i++;
      k++;
    }

    while (j < n2) {
      in[k] = R[j];
      j++;
      k++;
    }
  }
}

File: sorting.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void insertionSort(int arr[], int n);
void mergeSort(int arr[], int l, int r);
void merge(int arr[], int l, int m, int r);
void quickSort(int arr[], int l, int r);
int *createRandomArray(int size);
double time_insertion_sort(int size);
double time_merge_sort(int size);
double time_quick_sort(int size);
void printArray(int arr[], int n);

int main()
{
  srand(time(NULL));

  // Size of arrays for the initial tests
  // int size = 2000000;   // roughly max (in millions) that ubuntu machine is handling for some reason.
  int size = 2147483647;   // works on insertion, doesn't on the other ones interestingly

  // Single Runs:
  printf("Insertion Sort: %.6f seconds\n", time_insertion_sort(size));
  printf("Merge Sort: %.6f seconds\n", time_merge_sort(size));
  printf("Quick Sort: %.6f seconds\n", time_quick_sort(size));

  // Multiple Runs for Mean:
  // double insert_total = 0;
  // double merge_total = 0;
  // double quick_total = 0;
  // for(int i=0; i<20; i++)
  // {
  //   // insert_total += time_insertion_sort(size);
  //   // printf("insertion: %d\n", i);
  //   merge_total += time_merge_sort(size);
  //   printf("merge: %d\n", i);
  //   quick_total += time_quick_sort(size);
  //   printf("quick: %d\n", i);
  // }

  // printf("Insertion Sort [MEAN]: %.6f seconds\n", insert_total/20);
  // printf("Merge Sort [MEAN]: %.6f seconds\n", merge_total/20);
  // printf("Quick Sort [MEAN]: %.6f seconds\n", quick_total/20);

  return 0;
}

double time_merge_sort(int size)
{
  int *randomArray = createRandomArray(size);

  clock_t start_time = clock();
  mergeSort(randomArray, 0, size - 1);
  clock_t end_time = clock();

  // printArray(randomArray, 25);   // check things are being sorted properly

  // Free memory allocated for array
  free(randomArray);

  // Calculate time difference in seconds
  return ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
}

double time_quick_sort(int size)
{
  int *randomArray = createRandomArray(size);

  clock_t start_time = clock();
  quickSort(randomArray, 0, size - 1);
  clock_t end_time = clock();

  // printArray(randomArray, 25);   // check things are being sorted properly

  // Free memory allocated for array
  free(randomArray);

  // Calculate time difference in seconds
  return ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
}

double time_insertion_sort(int size)
{
  int *randomArray = createRandomArray(size);

  clock_t start_time = clock();
  insertionSort(randomArray, size);
  clock_t end_time = clock();

  // Free memory allocated for array
  free(randomArray);

  // Calculate time difference in seconds
  return ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
}

int *createRandomArray(int size)
{
  // printf("Creating random array of size: %d\n", size);
  int *randomArray = (int *)malloc(size * sizeof(int));
  if (randomArray == NULL)
  {
    fprintf(stderr, "Memory allocation failed\n");
    exit(EXIT_FAILURE);
  }

  // printf("-------------------------------------------------\n");
  // printf("Created random array of size: %d\n", size);
  // printf("Array is at memory address: %p\n", &size);
  // printf("Array is taking up %.4f GB of memory\n", (double)size * sizeof(int) / 1000000000);
  // printf("\n-------------------------------------------------\n");

  // printf("Filling array with random integers\n");
  for (int i = 0; i < size; i++)
  {
    randomArray[i] = rand(); // Fill with random integers
    if ((i + 1) % (size / 10) == 0)
    {
      // printf("Filled %.0f%% of array\n", (double)(i + 1) / size * 100);
    }
  }

  return randomArray;
}

void insertionSort(int arr[], int n)
{
  int i, key, j;
  for (i = 1; i < n; i++)
  {
    // printf(":: %d / %d ::\n", i, n);
    key = arr[i];
    j = i - 1;
    while (j >= 0 && arr[j] > key)
    {
      arr[j + 1] = arr[j];
      j = j - 1;
    }
    arr[j + 1] = key;
  }
}

void merge(int arr[], int l, int m, int r)
{
  int i, j, k;
  int n1 = m - l + 1;
  int n2 = r - m;

  int L[n1], R[n2];

  for (i = 0; i < n1; i++)
    L[i] = arr[l + i];
  for (j = 0; j < n2; j++)
    R[j] = arr[m + 1 + j];

  i = 0;
  j = 0;
  k = l;
  while (i < n1 && j < n2)
  {
    if (L[i] <= R[j])
    {
      arr[k] = L[i];
      i++;
    }
    else
    {
      arr[k] = R[j];
      j++;
    }
    k++;
  }

  while (i < n1)
  {
    arr[k] = L[i];
    i++;
    k++;
  }

  while (j < n2)
  {
    arr[k] = R[j];
    j++;
    k++;
  }
}

void mergeSort(int arr[], int l, int r)
{
  if (l < r)
  {
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);

    merge(arr, l, m, r);
  }
}

int partition(int arr[], int l, int r)
{
  int pivot = arr[r];
  int i = (l - 1);

  for (int j = l; j <= r - 1; j++)
  {
    if (arr[j] < pivot)
    {
      i++;
      int temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
    }
  }
  int temp = arr[i + 1];
  arr[i + 1] = arr[r];
  arr[r] = temp;
  return (i + 1);
}

void quickSort(int arr[], int l, int r)
{
  if (l < r)
  {
    int pi = partition(arr, l, r);
    quickSort(arr, l, pi - 1);
    quickSort(arr, pi + 1, r);
  }
}

void printArray(int arr[], int n)
{
  printf("[");
  for (int i = 0; i < n; i++)
  {
    printf("%d", arr[i]);
    if (i < n - 1)
    {
      printf(", ");
    }
  }
  printf("]\n");
}
